### PR TITLE
enable grain in input pipeline

### PR DESCRIPTION
**Pull request by aireenmei**
Convergence tested: loss 12.4 -> 3 (https://pantheon.corp.google.com/kubernetes/service/us-east5/v5e-256-bodaborg/default/aireen-v5e256-0129-0747/logs?e=13802955&mods=allow_workbench_image_override&project=tpu-prod-env-multipod)
checkpointing with data iterator will be in a separate PR

### CODE REVIEW

File: MaxText/train.py [hunk 1]
```diff
@@ -84,13 +85,18 @@ def get_first_step(state):
...
   """Loads the next batch. Can keep reusing the same batch for performance reasons """
 
   if config.reuse_example_batch and example_batch is not None:
     return example_batch
   else:
-    return train_iter()
+    if config.dataset_type == 'c4':
+      return train_iter()
+    elif config.dataset_type == 'c4-array_record':
+      return get_next_batch_sharded(train_iter, mesh)
```
Comment (rwitten) [inline MaxText/train.py:97]: I feel like this interface should be unified -- this code shouldn't know about dataset_type

File: end_to_end/test_convergence_1b_params_grain.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
```
Comment (rwitten) [inline end_to_end/test_convergence_1b_params_grain.sh:1]: Let's avoid duplication -- the other one can do this via overrides, right?

File: requirements.txt [hunk 3]
```diff
@@ -1,8 +1,11 @@
 orbax-checkpoint
 absl-py
+array-record
+argparse
```
Comment (rwitten) [inline requirements.txt:4]: Are you sure you use this?

**Review by rwitten — CHANGES_REQUESTED** [review]
Very close, nice change!

(Have you converted the full Grain dataset? How did that go?)

File: MaxText/train.py (hunk 1 above)
Comment (aireenmei) [reply MaxText/train.py]: revised as suggested.

File: end_to_end/test_convergence_1b_params_grain.sh (hunk 2 above)
Comment (aireenmei) [reply end_to_end/test_convergence_1b_params_grain.sh]: done

File: requirements.txt (hunk 3 above)
Comment (aireenmei) [reply requirements.txt]: Removed, thanks!

**Review by rwitten — APPROVED** [review]
Amazing!

**Pull request closed without merging**